spark-jobs/src/main/scala/filodb/downsampler/chunk/BatchExporter.scala

```diff
@@ -0,0 +1,223 @@
+   */
+  private def getTimeValuePairs(partKeyMap: Map[String, String],
+                                partition: ReadablePartition,
+                                chunkReader: ChunkSetInfoReader,
+                                irowStart: Int,
+                                irowEnd: Int): Iterator[(Long, Double)] = {
+    val timestampCol = 0  // FIXME: need a more dynamic (but efficient) solution
+    val columns = partition.schema.data.columns
+    val valueCol = columns.indexWhere(_.name == partition.schema.data.valueColName)
+    val timestampIter = getChunkColIter(partition, chunkReader, timestampCol, irowStart).asLongIt
+    columns(valueCol).columnType match {
+      case DoubleColumn =>
+        val valueIter = getChunkColIter(partition, chunkReader, valueCol, irowStart).asDoubleIt
+        (irowStart until irowEnd).iterator.map{ _ =>
+          (timestampIter.next, valueIter.next)
+        }
+      case colType =>
+        DownsamplerContext.dsLogger.warn(
+          s"BatchExporter encountered unhandled ColumnType=$colType; partKeyMap=$partKeyMap")
+        Iterator.empty
+    }
+  }
+
+  /**
+   * Returns true iff the label map matches all filters.
+   */
+  private def matchAllFilters(filters: Seq[ColumnFilter],
+                              labels: Map[String, String]): Boolean = {
+    filters.forall( filt =>
+      labels.get(filt.column)
+        .exists(value => filt.filter.filterFunc(value))
+    )
+  }
+
+  /**
+   * Returns the Spark Rows to be exported.
+   */
+  def getExportRows(readablePartitions: Seq[ReadablePartition],
+                    userStartTime: Long,
+                    userEndTime: Long): Iterator[Row] = {
+    val startMs = System.currentTimeMillis()
+    val rowIter = readablePartitions
+      .iterator
+      .map { part =>
+        // pair each partition with a map of its label-value pairs
+        val partKeyMap = {
+          val rawSchemaId = RecordSchema.schemaID(part.partKeyBytes, UnsafeUtils.arayOffset)
+          val schema = schemas(rawSchemaId)
+          schema.partKeySchema.toStringPairs(part.partKeyBytes, UnsafeUtils.arayOffset).toMap
+        }
+        (part, partKeyMap)
+      }
+      .filter { case (part, partKeyMap) => shouldExport(partKeyMap)}
+      .flatMap {case (part, partKeyMap) => getExportDatas(part, partKeyMap, userStartTime, userEndTime)}
+      .map(exportDataToRow)
+    val endMs = System.currentTimeMillis()
+    exportPrepBatchLatency.record(endMs - startMs)
+    rowIter
+  }
+
+  /**
+   * Converts an ExportData to a Spark Row.
+   * The result Row *must* match this.exportSchema.
+   */
+  private def exportDataToRow(exportData: ExportRowData): Row = {
+    val dataSeq = new mutable.ArrayBuffer[Any](3 + downsamplerSettings.exportPathSpecPairs.size)
+    dataSeq.append(
+      exportData.partKeyString,
+      exportData.timestamp,
+      exportData.value
+    )
+    dataSeq.appendAll(exportData.partitionStrings)
+    Row.fromSeq(dataSeq)
```

Comment:
I should add a `NOTE` here: immutable `Seq`s were a huge headache. When I'd just do something like:
```
Seq(val, val, val) ++ exportData.partitionStrings.toSeq
```
I'd get `ArrayIndexOutOfBounds` exceptions during the export, when `Row`s were compared against the `DataFrame` schema. This saves us two allocations, I guess 🤷

Reply:
Note added.